docs/proposals/scheduling/20220629-fine-grained-device-scheduling.md

```diff
@@ -0,0 +1,419 @@
+    // Health indicates whether the device is normal
+    Health bool `json:"health,omitempty"`
+    // Resources represents the total capacity of various resources of the device
+    Resources map[string]resource.Quantity `json:"resource,omitempty"`
+}
+
+type DevicesStatus struct {
+    Allocations []DeviceAllocation `json:"allocations"`  // record each pod device usage card list, more detail see Compatibility    
+}
+
+type DeviceAllocation struct {
+    Type    DeviceType             `json:"type"`
+    Entries []DeviceAllocationItem `json:"entries"`
+}
+
+type DeviceAllocationItem struct {
+    Name      string   `json:"name"`
+    Namespace string   `json:"namespace"`
+    UUID      string   `json:"uuid"`
+    Devices   []string `json:"devices"`
+}
+```
+
+#### koordlet and koord-runtime-proxy
+
+Our target is to work compatible with origin k8s kubelet and k8s device plugins, so:
+1. We still allow kubelet and device plugin to allocate concrete device, which means no matter there's a k8s device 
+plugin or not, our design can work well.
+2. In koord-runtime-proxy, we will use Pod's `DeviceAllocation` in annotation to replace the step1's result of container's 
+args and envs.
+
+We should modify protocol between koord-runtime-proxy and koordlet to add container env:
+
+```go
+type ContainerResourceHookRequest struct {  
+    ....
+    Env map[string]string
+}
+
+type ContainerResourceHookResponse struct {
+    ....
+    Env map[string]string
+}
+```
+
+Then we will add a new `gpu-hook` in koordlet's runtimehooks. It's used in `PreStartContainer` and `PostContainerstage`. 
+We will generate new GPU env `NVIDIA_VISIBLE_DEVICES` by Pod GPU allocation result in annotation. 
+
+When we handle hot-update processing, we can handle the existing scheduled Pods without device allocation in pod's annotation. If GPU allocation info is not in annotation, we will find the GPU allocations from `ContainerResourceHookRequest`'s `Env`, and we will update all GPU allocations to Device CRD instance. 
+
+The koord-runtime-proxy can see these Pod's env, we need koord-runtime-proxy to pass these environments to koordlet, and koordlet parse the GPU related env to find the concrete device ids. Then koordlet will update these information to Device CRD intance, which help scheduler to recover the deviceCache.
+
+Besides, the koordlet should report GPU model to node labels same as device plugin, this is in-case koordinator working without device-plugin.
+
+Finally, we should modify `ContainerResourceExecutor`'s `UpdateRequest` function in koord-runtime-proxy, and let new GPU env covering old GPU env.
+
+### Compatibility
+
+As we know, the GPU scheduling in kube-scheduler side has no any different with other scalar resources. The concrete 
+device-level assigning is done by kubelet and GPU device plugin, which will generate container's GPU env. 
+
+Our design has no conflict with the above process. our device reporter will report koordinator GPU resources for kubelet
```

Comment:
> our device reporter will report koordinator GPU resources for kubelet updating node resources

how does device reporter report koordinator GPU resources to kubelet? 

does this mean that device reporter still implements a device plugin for koordinator GPU resources?

Reply:
resolved after reading https://github.com/koordinator-sh/koordinator/pull/410